// markup-it/lib/markdown/re/table.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _utils = require('../utils');

var _inline = require('./inline');

var _inline2 = _interopRequireDefault(_inline);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var pipe = /\|/;

var table = {
    cellSeparation: /^pipe/,

    nptable: /^ *(\S.*\|.*)\n *([-:]+ *\|[-| :]*)\n((?:.*\|.*(?:\n|$))*)\n*/,
    normal: /^ *\|(.+)\n *\|( *[-:]+[-| :]*)\n((?: *\|.*(?:\n|$))*)\n*/,

    // Split a row into cells
    cell: / *pipe */,

    // Replace trailing pipe
    trailingPipe: /^ *| *pipe *$/g,

    // Remove trailing pipe of align
    trailingPipeAlign: /^ *|pipe *$/g,

    // Remove trailing pipe of cell
    trailingPipeCell: /(?: *pipe *)?\n$/,

    // Remove edge pipes of a cell
    edgePipesCell: /^ *pipe *| *pipe *$/g,

    // Alignements
    alignRight: /^ *-+: *$/,
    alignCenter: /^ *:-+: *$/,
    alignLeft: /^ *:-+ *$/
};

table.cellSeparation = (0, _utils.replace)(table.cellSeparation)(/pipe/g, pipe)();
table.cell = (0, _utils.replace)(table.cell)(/pipe/g, pipe)();
table.trailingPipe = (0, _utils.replace)(table.trailingPipe, 'g')(/pipe/g, pipe)();
table.trailingPipeAlign = (0, _utils.replace)(table.trailingPipeAlign, 'g')(/pipe/g, pipe)();
table.trailingPipeCell = (0, _utils.replace)(table.trailingPipeCell, 'g')(/pipe/g, pipe)();
table.edgePipesCell = (0, _utils.replace)(table.edgePipesCell, 'g')(/pipe/g, pipe)();
table.cellInlineText = (0, _utils.replace)(_inline2.default.text)(']|', '|]|')();
table.cellInlineEscape = _inline2.default.escape;

exports.default = table;
