// markup-it/lib/markdown/inlines/image.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _immutable = require('immutable');

var _ = require('../../');

var _inline = require('../re/inline');

var _inline2 = _interopRequireDefault(_inline);

var _utils = require('../utils');

var utils = _interopRequireWildcard(_utils);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Resolve an image reference
 * @param  {State} state
 * @param  {String} refID
 * @return {Map} data?
 */
function resolveImageRef(state, refID) {
    var data = utils.resolveRef(state, refID);

    if (!data) {
        return null;
    }

    return data.set('src', data.get('href')).remove('href');
}

/**
 * Test if a link input is an image
 * @param {String} raw
 * @return {Boolean}
 */
function isImage(raw) {
    return raw.charAt(0) === '!';
}

/**
 * Serialize a image to markdown
 * @type {Serializer}
 */
var serialize = (0, _.Serializer)().matchType(_.INLINES.IMAGE).then(function (state) {
    var node = state.peek();
    var data = node.data;

    // Escape the url

    var src = utils.escapeURL(data.get('src') || '');

    var alt = utils.escape(data.get('alt') || '');
    var title = utils.escape(data.get('title') || '');

    var output = void 0;

    if (title) {
        output = '![' + alt + '](' + src + ' "' + title + '")';
    } else {
        output = '![' + alt + '](' + src + ')';
    }

    return state.shift().write(output);
});

/**
 * Deserialize a classic image like:
 *  ![Hello](test.png)
 * @type {Deserializer}
 */
var deserializeNormal = (0, _.Deserializer)().matchRegExp(_inline2.default.link, function (state, match) {
    if (!isImage(match[0])) {
        return undefined;
    }

    var data = (0, _immutable.Map)({
        alt: match[1] ? utils.unescape(match[1]) : undefined,
        src: utils.unescapeURL(match[2]),
        title: match[3] ? utils.unescape(match[3]) : undefined
    }).filter(Boolean);

    var node = _.Inline.create({
        type: _.INLINES.IMAGE,
        isVoid: true,
        data: data
    });

    return state.push(node);
});

/**
 * Deserialize a reference image:
 *  nolink: ![1]
 * @type {Deserializer}
 */
var deserializeRef = (0, _.Deserializer)().matchRegExp([_inline2.default.reflink, _inline2.default.nolink], function (state, match) {
    if (!isImage(match[0])) {
        return undefined;
    }

    var refID = match[2] || match[1];
    var data = resolveImageRef(state, refID);

    if (!data) {
        return undefined;
    }

    var node = _.Inline.create({
        type: _.INLINES.IMAGE,
        isVoid: true,
        data: data
    });

    return state.push(node);
});

var deserialize = (0, _.Deserializer)().use([deserializeNormal, deserializeRef]);

exports.default = { serialize: serialize, deserialize: deserialize };
