// react-rich-diff/lib/components/ChangeIndicator.js
'use strict';

var React = require('react');

/**
 * Render an indicator for a change (added, modified, etc) with the right class name.
 * @type {ReactClass}
 */
var ChangeIndicator = React.createClass({
    displayName: 'ChangeIndicator',

    propTypes: {
        kind: React.PropTypes.string.isRequired,
        type: React.PropTypes.string.isRequired,
        children: React.PropTypes.node
    },

    render: function render() {
        var _props = this.props,
            kind = _props.kind,
            type = _props.type,
            children = _props.children;

        var className = 'diff-' + kind + '-' + type;

        if (kind == 'block') {
            return React.createElement(
                'div',
                { className: className },
                children
            );
        } else {
            return React.createElement(
                'span',
                { className: className },
                children
            );
        }
    }
});

module.exports = ChangeIndicator;
