// react-rich-diff/lib/components/RichDiff.js
'use strict';

var React = require('react');
var Changes = require('./Changes');

/**
 * Render an entire diff from a state.
 * @type {React}
 */
var RichDiff = React.createClass({
    displayName: 'RichDiff',

    propTypes: {
        className: React.PropTypes.string,
        state: React.PropTypes.object.isRequired,
        minToWrap: React.PropTypes.number
    },

    getDefaultProps: function getDefaultProps() {
        return {
            className: '',
            minToWrap: 3
        };
    },
    render: function render() {
        var _props = this.props,
            state = _props.state,
            className = _props.className,
            minToWrap = _props.minToWrap;


        return React.createElement(Changes, {
            Wrapper: function Wrapper(props) {
                return React.createElement(
                    'div',
                    { className: 'RichDiff ' + className },
                    props.children
                );
            },
            changes: state.changes,
            minToWrap: minToWrap
        });
    }
});

module.exports = RichDiff;
